keel-plugin/src/main/kotlin/com/netflix/spinnaker/keel/plugin/ResourceHandler.kt

```diff
@@ -12,78 +12,86 @@ import com.netflix.spinnaker.keel.api.randomUID
 import com.netflix.spinnaker.keel.diff.ResourceDiff
 import com.netflix.spinnaker.keel.events.Task
 import org.slf4j.Logger
+import org.slf4j.LoggerFactory
 
 /**
  * @param S the spec type.
  * @param R the resolved model type.
  *
  * If those two are the same, use [SimpleResourceHandler] instead.
  */
-interface ResourceHandler<S : ResourceSpec, R : Any> : KeelPlugin {
+abstract class ResourceHandler<S : ResourceSpec, R : Any>(
+  private val objectMapper: ObjectMapper,
+  private val resolvers: List<Resolver<*>>
+) : KeelPlugin {
 
-  val log: Logger
+  protected val log: Logger by lazy { LoggerFactory.getLogger(javaClass) }
 
-  val objectMapper: ObjectMapper
-
-  val normalizers: List<ResourceNormalizer<*>>
-
-  val apiVersion: ApiVersion
+  abstract val apiVersion: ApiVersion
 
   /**
    * Maps the kind to the implementation type.
    */
-  val supportedKind: Pair<ResourceKind, Class<S>>
+  abstract val supportedKind: Pair<ResourceKind, Class<S>>
 
   /**
-   * Validates the resource spec, and generates a metadata header, and applies any defaults /
-   * normalization.
+   * Validates the resource spec and generates a metadata header.
    *
-   * Implementors should not generally need to override this method,
-   *  custom normalization and validation should be done with a [ResourceNormalizer].
+   * Implementors should not generally need to override this method.
    *
-   * @return a hydrated `Resource` with a name generated by convention, a `uid`, default values
-   * applied, etc.
+   * @return a hydrated `Resource` with a name generated by convention, an `id`,  `uid`, etc.
    */
   fun normalize(resource: SubmittedResource<S>): Resource<S> {
     val metadata = resource.metadata + mapOf(
       "id" to resource.id.toString(),
       "uid" to randomUID().toString(),
       "application" to resource.spec.application
     )
-    val hydratedResource = Resource(
+    return Resource(
       apiVersion = resource.apiVersion,
       kind = resource.kind,
       metadata = metadata,
       spec = resource.spec
     )
-    return normalize(hydratedResource)
```

Comment:
Normalizing doesn't call the resolvers any more. Just creates required metadata.